File: Guidance Packages/SmartClientFactory/SmartClientFactoryPackage/Actions/Utility.cs
//===============================================================================
// Microsoft patterns & practices
// Smart Client Software Factory 2010
//===============================================================================
// Copyright (c) Microsoft Corporation.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
//===============================================================================
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
//===============================================================================
using System;
using System.Collections.Generic;
using System.Text;
using EnvDTE;
using System.ComponentModel;
using Microsoft.Practices.ComponentModel;
using System.Text.RegularExpressions;
using System.Globalization;
using Microsoft.VisualStudio.Shell.Interop;
using Microsoft.Practices.RecipeFramework.Extensions.CommonHelpers;

namespace Microsoft.Practices.SmartClientFactory.Actions
{
	class Utility
	{
		private const uint ROOT = 0xFFFFFFFE;
		
		internal static Guid GetProjectGuid(IServiceProvider serviceProvider, Project project)
		{
			if (project != null)
			{
				IVsHierarchy vsHier = Microsoft.Practices.RecipeFramework.Library.DteHelper.GetVsHierarchy(serviceProvider, project);
				Guid projectGuid = Guid.Empty;
				vsHier.GetGuidProperty(ROOT, (int)__VSHPROPID.VSHPROPID_ProjectIDGuid, out projectGuid);
				return projectGuid;
			}
			return Guid.Empty;
		}


		internal static Project GetProjectFromGuid(DTE dte, IServiceProvider provider, Guid guid)
		{
			return DteHelperEx.FindProject(dte, delegate(Project match)
			{
				return guid == GetProjectGuid(provider, match);
			});
		}

		internal static bool IsSealedOrStatic(CodeClass codeClass)
		{
			EditPoint start = codeClass.StartPoint.CreateEditPoint();
			EditPoint cursor = codeClass.StartPoint.CreateEditPoint();
			cursor.WordRight(1);
			string currentWord = String.Empty;
			while (!cursor.AtEndOfDocument)
			{
				currentWord = start.GetText(cursor);
				if (currentWord.StartsWith("class", true, CultureInfo.InvariantCulture)) return false;
				if (currentWord.StartsWith("sealed", true, CultureInfo.InvariantCulture)) return true;
				if (currentWord.StartsWith("static", true, CultureInfo.InvariantCulture)) return true;
                if (currentWord.StartsWith("notinheritable", true, CultureInfo.InvariantCulture)) return true;
				start.WordRight(1);
				cursor.WordRight(1);
			}
			return false;
		}
	}
}

File: Blocks/CAB/VB/UnitTests/CompositeUI.WinForms/Utility.vb
'===============================================================================
' Microsoft patterns & practices
' CompositeUI Application Block
'===============================================================================
' Copyright © Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================

Imports Microsoft.VisualBasic


Imports Microsoft.VisualStudio.TestTools.UnitTesting








Imports System
Imports Microsoft.Practices.CompositeUI.SmartParts

Public Class Utility
	Public Sub EventHandler(ByVal sender As Object, ByVal args As WorkspaceCancelEventArgs)
		args.Cancel = True
	End Sub
End Class

Public Class FiresSmartPartUtility(Of TEventArgs As WorkspaceEventArgs)
	Private innerSmartPart As Object

	Public Property SmartPart() As Object
		Get
			Return innerSmartPart
		End Get
		Set(ByVal value As Object)
			innerSmartPart = value
		End Set
	End Property

	Public Sub New(ByVal smartPart As Object)
		innerSmartPart = smartPart
	End Sub

	Public Sub EventHandler(ByVal sender As Object, ByVal args As TEventArgs)
		innerSmartPart = args.SmartPart
	End Sub
End Class

Public Class FiresCounterUtility(Of TEventArgs As WorkspaceEventArgs)

	Protected innerActivated As Integer

	Public Property Activated() As Integer
		Get
			Return innerActivated
		End Get
		Set(ByVal value As Integer)
			innerActivated = value
		End Set
	End Property

	Public Sub New(ByVal activated As Integer)
		innerActivated = activated
	End Sub

	Public Sub EventHandler(ByVal sender As Object, ByVal args As TEventArgs)
		innerActivated += 1
	End Sub

End Class

Public Class FiresCounterWithSmartPartCheckUtility(Of TEventArgs As WorkspaceEventArgs)
	Inherits FiresCounterUtility(Of TEventArgs)
	Private innerSmartPart As Object
	Public Property SmartPart() As Object
		Get
			Return innerSmartPart
		End Get
		Set(ByVal value As Object)
			innerSmartPart = value
		End Set
	End Property
	Public Sub New(ByVal activated As Integer, ByVal smartPart As Object)
		MyBase.new(activated)
		Me.innerSmartPart = smartPart
	End Sub
	Public Overloads Sub EventHandler(ByVal sender As Object, ByVal args As TEventArgs)
		innerActivated += 1
		Assert.AreSame(args.SmartPart, innerSmartPart)
	End Sub
End Class

Public Class StateChangedUtility(Of TEventArgs As EventArgs)

	Protected innerStateChanged As Boolean

	Public Property StateChanged() As Boolean
		Get
			Return innerStateChanged
		End Get
		Set(ByVal value As Boolean)
			innerStateChanged = value
		End Set
	End Property

	Public Sub New(ByVal stateChanged As Boolean)
		innerStateChanged = stateChanged
	End Sub

	Public Sub EventHandler(ByVal sender As Object, ByVal args As TEventArgs)
		innerStateChanged = True
	End Sub

End Class

Public Class StateChangedWithSmartPartCheckUtility(Of TEventArgs As WorkspaceEventArgs)
	Inherits StateChangedUtility(Of TEventArgs)

	Private innerSmartPart As Object
	Public Property SmartPart() As Object
		Get
			Return innerSmartPart
		End Get
		Set(ByVal value As Object)
			innerSmartPart = value
		End Set
	End Property
	Public Sub New(ByVal stateChanged As Boolean)
		MyBase.New(stateChanged)
		innerSmartPart = Nothing
	End Sub

	Public Overloads Sub EventHandler(ByVal sender As Object, ByVal args As TEventArgs)
		innerStateChanged = True
		innerSmartPart = args.SmartPart
	End Sub

End Class